aeron-archive/src/main/java/io/aeron/archive/ArchiveConductor.java

```diff
@@ -1328,54 +1329,86 @@ void migrateSegments(
-            final ArrayDeque<String> files = new ArrayDeque<>();
+            final RecordingSummary srcRecordingSummary =
+                catalog.recordingSummary(srcRecordingId, recordingSummary);
+
+            final RecordingSummary dstRecordingSummary =
+                catalog.recordingSummary(dstRecordingId, otherRecordingSummary);
 
-            while (position >= segmentFileBasePosition)
+            if (isActiveRecording(controlSession, correlationId, recordingSummary) ||
+                !hasMatchingStreamParameters(controlSession, correlationId, srcRecordingSummary, dstRecordingSummary))
             {
-                final String segmentFileName = segmentFileName(srcRecordingId, position);
-                final File srcFile = new File(archiveDir, segmentFileName);
-                if (position == recordingSummary.stopPosition)
-                {
-                    files.addFirst(segmentFileName);
-                    position -= segmentLength;
-                    continue;
-                }
-                else if (!srcFile.exists())
-                {
-                    break;
-                }
+                return;
+            }
 
-                final String dstFile = segmentFileName(dstRecordingId, position);
-                if (!srcFile.renameTo(new File(archiveDir, dstFile)))
-                {
-                    final String msg = "failed to rename " + srcFile + " to " + dstFile;
-                    controlSession.sendErrorResponse(correlationId, msg, controlResponseProxy);
-                    return;
-                }
+            final boolean canPrepend = srcRecordingSummary.stopPosition == dstRecordingSummary.startPosition;
+            final boolean canAppend = srcRecordingSummary.startPosition == dstRecordingSummary.stopPosition;
 
-                attachedSegmentCount++;
-                position -= segmentLength;
+            final long seamPosition;
+
+            if (canPrepend)
+            {
+                seamPosition = srcRecordingSummary.stopPosition;
+            }
+            else if (canAppend)
+            {
+                seamPosition = srcRecordingSummary.startPosition;
+            }
+            else
+            {
+                final String msg = "invalid migrate: src and dst are not contiguous" +
+                    " srcStartPosition=" + srcRecordingSummary.startPosition +
+                    " srcStopPosition=" + srcRecordingSummary.stopPosition +
+                    " dstStartPosition=" + dstRecordingSummary.startPosition +
+                    " dstStopPosition=" + dstRecordingSummary.stopPosition;
+                controlSession.sendErrorResponse(correlationId, msg, controlResponseProxy);
+                return;
+            }
+
+            if (isSeamPosSegmentUnaligned(controlSession, correlationId, "src", srcRecordingSummary, seamPosition) ||
+                isSeamPosSegmentUnaligned(controlSession, correlationId, "dst", dstRecordingSummary, seamPosition))
```

Comment:
I prefer `!positivePredicate` rather than `negativePredicate`, but the latter makes IntelliJ happy without suppressing warnings.